modules/programs/hexchat.nix

```diff
@@ -0,0 +1,256 @@
+    };
+    enable = mkEnableOption "HexChat - Graphical IRC client";
+    theme = mkOption {
+      default = null;
+      description = "Theme package for HexChat.";
+      # example = ""; # To-Do: Update here with a valid theme.
+      type = with lib.types; nullOr package;
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    home.packages = [ pkgs.hexchat ];
+    xdg.configFile."hexchat" = lib.mkIf (cfg.theme != null) {
+      source = cfg.theme.out;
+      recursive = true;
+    };
+    xdg.configFile."hexchat/hexchat.conf" = lib.mkIf (cfg.conf != null) {
+      force = true;
+      text = let hexchatConf = cfg.conf;
+      in concatMapStringsSep "\n" (x: x + " = " + hexchatConf.${x})
+      (attrNames hexchatConf);
+    };
+    xdg.configFile."hexchat/servlist.conf" = lib.mkIf (cfg.channels != null)
+      (if attrNames cfg.channels == [ ] then {
+        text = "";
+      } else {
+        force = true;
+        text = let
+          tr_channel = (channel_name:
+            let
+              channel = cfg.channels.${channel_name};
+              listChar = (c: l:
+                (concatMapStrings (item: ''
+
+                  ${c}=${item}'') l));
+            in let
+              name = ''
+
+                N='' + channel_name;
+              nickname = ''
+
+                I='' + channel.nickname;
+              nickname2 = ''
+
+                i='' + channel.nickname2;
+              real_name = ''
+
+                R='' + channel.real_name;
+              user_name = ''
+
+                U='' + channel.user_name;
+              password = ''
+
+                P='' + channel.password;
+              charset = ''
+
+                E='' + channel.charset;
+              login_method = ''
+
+                L='' + toString channel.login_method;
+              servers = listChar "S" channel.servers;
+              autojoin = listChar "J" channel.autojoin;
+              commands = listChar "C" channel.commands;
+              options = with channel.options;
+                (''
+
+                  F='' + toString (if channel.options == null then
+                    0
+                  else
+                    ((if autoconnect then 8 else 0)
+                      + (if !connect_to_selected_server_only then 1 else 0)
+                      + (if !bypass_proxy then 16 else 0)
+                      + (if use_ssl_for_all_servers then 4 else 0)
+                      + (if accept_invalid_ssl_certificates then 32 else 0)
+                      + (if use_global_user_information then 2 else 0))));
+              # Note: Missing option `D=`.
+            in name + login_method + charset + options +
+            # Note: Position for `D=`.
+            nickname + nickname2 + real_name + user_name + password + servers
+            + autojoin + commands);
+        in let
+          channels =
+            concatMapStringsSep "\n" tr_channel (attrNames cfg.channels);
+        in ''
+          v=2.14.3
+        '' + channels + "\n\n";
```

Comment:
All those  `let` inside another `let` makes the code really difficult to read.

Reply:
What made this look ugly was `nixfmt`. Which converted "\n" to ''<line break>''. On the odd line breaking pattern it seems there is not much that can be done because HexChat requires line breaks for each field as it is there.

Regarding the let. There is a logic to it. Nixfmt is what has hidden it. I'm not sure how it could be improved.

What could be improved on is naming. Like: "tr_channel" could be `transformChannel`?

Reply:
Yeah, now that I am reading with more attention it seems the way the code is currently written it needs some of those `let` (I don't think all of them are needed, but it is difficult to understand even formatting the code correctly).

Moving this to smaller functions will probably help, and also using string interpolation instead of concat. Another suggestion is maybe doing something like this:

```
channelConfig = ''
  N=${config.channels.${channel_name}}
  I=${channel.nickname}
  ...
'';
```

This will make it easier because IMO, it will make it more clear what is the final format of the config file. You can still do some of the logic inside `${}` block (but the logic on `F=` maybe would be better moved to a function since it would be too big otherwise).

Reply:
I'm already changing this at this moment. I'll push soon.

Reply:
Also, as I explained it [here](https://github.com/nix-community/home-manager/pull/1724#discussion_r562561774) maybe it would be better to move most of this let to a function called `mkConfig` or something defined in the `let` in line 4.

This will also help reducing the number of indentations, making `nixfmt` behave better.